File: exercises/2-intermediate/config/webapp-controller/main.go
package main

import (
    "context"
    "flag"
    "fmt"
    "os"
    "path/filepath"
    
    appsv1 "k8s.io/api/apps/v1"
    corev1 "k8s.io/api/core/v1"
    networkingv1 "k8s.io/api/networking/v1"
    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
    "k8s.io/apimachinery/pkg/runtime"
    "k8s.io/apimachinery/pkg/util/intstr"
    "k8s.io/client-go/kubernetes"
    "k8s.io/client-go/tools/clientcmd"
    "k8s.io/client-go/util/homedir"
)

type WebApp struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    Spec             WebAppSpec   `json:"spec"`
}

type WebAppSpec struct {
    Image     string             `json:"image"`
    Port      int32             `json:"port"`
    Replicas  int32             `json:"replicas"`
    Domains   []string          `json:"domains,omitempty"`
    SSL       *SSLConfig        `json:"ssl,omitempty"`
    Resources *ResourceRequests `json:"resources,omitempty"`
}

type SSLConfig struct {
    Enabled    bool   `json:"enabled"`
    SecretName string `json:"secretName,omitempty"`
}

type ResourceRequests struct {
    Limits   *Resources `json:"limits,omitempty"`
    Requests *Resources `json:"requests,omitempty"`
}

type Resources struct {
    CPU    string `json:"cpu,omitempty"`
    Memory string `json:"memory,omitempty"`
}

func main() {
    kubeconfig := flag.String("kubeconfig", filepath.Join(homedir.HomeDir(), ".kube", "config"), "kubeconfig file")
    flag.Parse()

    config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
    if err != nil {
        fmt.Printf("Error building config: %v\n", err)
        os.Exit(1)
    }

    clientset, err := kubernetes.NewForConfig(config)
    if err != nil {
        fmt.Printf("Error creating clientset: %v\n", err)
        os.Exit(1)
    }

    // Example WebApp
    webapp := &WebApp{
        ObjectMeta: metav1.ObjectMeta{
            Name:      "example-webapp",
            Namespace: "default",
        },
        Spec: WebAppSpec{
            Image:    "nginx:1.14",
            Port:     80,
            Replicas: 3,
            SSL: &SSLConfig{
                Enabled:    true,
                SecretName: "webapp-tls",
            },
            Resources: &ResourceRequests{
                Limits: &Resources{
                    CPU:    "500m",
                    Memory: "512Mi",
                },
                Requests: &Resources{
                    CPU:    "250m",
                    Memory: "256Mi",
                },
            },
        },
    }

    // Create deployment
    deployment := createDeployment(webapp)
    _, err = clientset.AppsV1().Deployments(webapp.Namespace).Create(context.TODO(), deployment, metav1.CreateOptions{})
    if err != nil {
        fmt.Printf("Error creating deployment: %v\n", err)
    }

    // Create service
    service := createService(webapp)
    _, err = clientset.CoreV1().Services(webapp.Namespace).Create(context.TODO(), service, metav1.CreateOptions{})
    if err != nil {
        fmt.Printf("Error creating service: %v\n", err)
    }

    if len(webapp.Spec.Domains) > 0 && webapp.Spec.SSL != nil && webapp.Spec.SSL.Enabled {
        // Create Ingress with TLS
        ingress := createIngress(webapp)
        _, err = clientset.NetworkingV1().Ingresses(webapp.Namespace).Create(context.TODO(), ingress, metav1.CreateOptions{})
        if err != nil {
            fmt.Printf("Error creating ingress: %v\n", err)
        }
    }
}

func createDeployment(webapp *WebApp) *appsv1.Deployment {
    return &appsv1.Deployment{
        ObjectMeta: metav1.ObjectMeta{
            Name:      webapp.Name,
            Namespace: webapp.Namespace,
        },
        Spec: appsv1.DeploymentSpec{
            Replicas: &webapp.Spec.Replicas,
            Selector: &metav1.LabelSelector{
                MatchLabels: map[string]string{
                    "app": webapp.Name,
                },
            },
            Template: corev1.PodTemplateSpec{
                ObjectMeta: metav1.ObjectMeta{
                    Labels: map[string]string{
                        "app": webapp.Name,
                    },
                },
                Spec: corev1.PodSpec{
                    Containers: []corev1.Container{
                        {
                            Name:  webapp.Name,
                            Image: webapp.Spec.Image,
                            Ports: []corev1.ContainerPort{
                                {
                                    ContainerPort: webapp.Spec.Port,
                                },
                            },
                            Resources: createResourceRequirements(webapp.Spec.Resources),
                        },
                    },
                },
            },
        },
    }
}

func createService(webapp *WebApp) *corev1.Service {
    return &corev1.Service{
        ObjectMeta: metav1.ObjectMeta{
            Name:      webapp.Name,
            Namespace: webapp.Namespace,
        },
        Spec: corev1.ServiceSpec{
            Selector: map[string]string{
                "app": webapp.Name,
            },
            Ports: []corev1.ServicePort{
                {
                    Port:       webapp.Spec.Port,
                    TargetPort: intstr.FromInt(int(webapp.Spec.Port)),
                },
            },
        },
    }
}

func createIngress(webapp *WebApp) *networkingv1.Ingress {
    pathType := networkingv1.PathTypePrefix
    ingressRules := make([]networkingv1.IngressRule, 0, len(webapp.Spec.Domains))
    
    for _, domain := range webapp.Spec.Domains {
        ingressRules = append(ingressRules, networkingv1.IngressRule{
            Host: domain,
            IngressRuleValue: networkingv1.IngressRuleValue{
                HTTP: &networkingv1.HTTPIngressRuleValue{
                    Paths: []networkingv1.HTTPIngressPath{
                        {
                            Path:     "/",
                            PathType: &pathType,
                            Backend: networkingv1.IngressBackend{
                                Service: &networkingv1.IngressServiceBackend{
                                    Name: webapp.Name,
                                    Port: networkingv1.ServiceBackendPort{
                                        Number: webapp.Spec.Port,
                                    },
                                },
                            },
                        },
                    },
                },
            },
        })
    }

    var tls []networkingv1.IngressTLS
    if webapp.Spec.SSL != nil && webapp.Spec.SSL.Enabled {
        tls = []networkingv1.IngressTLS{
            {
                Hosts:      webapp.Spec.Domains,
                SecretName: webapp.Spec.SSL.SecretName,
            },
        }
    }

    return &networkingv1.Ingress{
        ObjectMeta: metav1.ObjectMeta{
            Name:      webapp.Name,
            Namespace: webapp.Namespace,
        },
        Spec: networkingv1.IngressSpec{
            TLS:   tls,
            Rules: ingressRules,
        },
    }
}

func createResourceRequirements(resources *ResourceRequests) corev1.ResourceRequirements {
    if resources == nil {
        return corev1.ResourceRequirements{}
    }

    reqs := corev1.ResourceRequirements{}
    
    if resources.Limits != nil {
        reqs.Limits = corev1.ResourceList{}
        if resources.Limits.CPU != "" {
            reqs.Limits[corev1.ResourceCPU] = resource.MustParse(resources.Limits.CPU)
        }
        if resources.Limits.Memory != "" {
            reqs.Limits[corev1.ResourceMemory] = resource.MustParse(resources.Limits.Memory)
        }
    }

    if resources.Requests != nil {
        reqs.Requests = corev1.ResourceList{}
        if resources.Requests.CPU != "" {
            reqs.Requests[corev1.ResourceCPU] = resource.MustParse(resources.Requests.CPU)
        }
        if resources.Requests.Memory != "" {
            reqs.Requests[corev1.ResourceMemory] = resource.MustParse(resources.Requests.Memory)
        }
    }

    return reqs
}

File: exercises/2-intermediate/config/webapp-controller/main.py
from dataclasses import dataclass
from typing import List, Optional
from kubernetes import client, config
from pathlib import Path
import argparse
import os

@dataclass
class Resources:
    cpu: Optional[str] = None
    memory: Optional[str] = None

@dataclass
class ResourceRequests:
    limits: Optional[Resources] = None
    requests: Optional[Resources] = None

@dataclass
class SSLConfig:
    enabled: bool = False
    secret_name: Optional[str] = None

@dataclass
class WebAppSpec:
    image: str
    port: int
    replicas: int
    domains: Optional[List[str]] = None
    ssl: Optional[SSLConfig] = None
    resources: Optional[ResourceRequests] = None

@dataclass
class WebApp:
    name: str
    namespace: str
    spec: WebAppSpec

def create_deployment(webapp: WebApp) -> client.V1Deployment:
    container = client.V1Container(
        name=webapp.name,
        image=webapp.spec.image,
        ports=[client.V1ContainerPort(container_port=webapp.spec.port)]
    )

    # Add resource requirements if specified
    if webapp.spec.resources:
        limits = {}
        requests = {}
        
        if webapp.spec.resources.limits:
            if webapp.spec.resources.limits.cpu:
                limits['cpu'] = webapp.spec.resources.limits.cpu
            if webapp.spec.resources.limits.memory:
                limits['memory'] = webapp.spec.resources.limits.memory
                
        if webapp.spec.resources.requests:
            if webapp.spec.resources.requests.cpu:
                requests['cpu'] = webapp.spec.resources.requests.cpu
            if webapp.spec.resources.requests.memory:
                requests['memory'] = webapp.spec.resources.requests.memory
                
        container.resources = client.V1ResourceRequirements(
            limits=limits,
            requests=requests
        )

    return client.V1Deployment(
        metadata=client.V1ObjectMeta(
            name=webapp.name,
            namespace=webapp.namespace
        ),
        spec=client.V1DeploymentSpec(
            replicas=webapp.spec.replicas,
            selector=client.V1LabelSelector(
                match_labels={"app": webapp.name}
            ),
            template=client.V1PodTemplateSpec(
                metadata=client.V1ObjectMeta(
                    labels={"app": webapp.name}
                ),
                spec=client.V1PodSpec(
                    containers=[container]
                )
            )
        )
    )

def create_service(webapp: WebApp) -> client.V1Service:
    return client.V1Service(
        metadata=client.V1ObjectMeta(
            name=webapp.name,
            namespace=webapp.namespace
        ),
        spec=client.V1ServiceSpec(
            selector={"app": webapp.name},
            ports=[client.V1ServicePort(
                port=webapp.spec.port,
                target_port=webapp.spec.port
            )]
        )
    )

def create_ingress(webapp: WebApp) -> client.V1Ingress:
    path_type = "Prefix"
    rules = []
    
    for domain in webapp.spec.domains:
        rules.append(client.V1IngressRule(
            host=domain,
            http=client.V1HTTPIngressRuleValue(
                paths=[client.V1HTTPIngressPath(
                    path="/",
                    path_type=path_type,
                    backend=client.V1IngressBackend(
                        service=client.V1IngressServiceBackend(
                            name=webapp.name,
                            port=client.V1ServiceBackendPort(
                                number=webapp.spec.port
                            )
                        )
                    )
                )]
            )
        ))

    tls = None
    if webapp.spec.ssl and webapp.spec.ssl.enabled:
        tls = [client.V1IngressTLS(
            hosts=webapp.spec.domains,
            secret_name=webapp.spec.ssl.secret_name
        )]

    return client.V1Ingress(
        metadata=client.V1ObjectMeta(
            name=webapp.name,
            namespace=webapp.namespace
        ),
        spec=client.V1IngressSpec(
            tls=tls,
            rules=rules
        )
    )

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--kubeconfig",
        default=os.path.join(str(Path.home()), ".kube", "config"),
        help="Path to kubeconfig file"
    )
    args = parser.parse_args()

    # Load kubernetes configuration
    config.load_kube_config(args.kubeconfig)

    # Create API clients
    apps_v1 = client.AppsV1Api()
    core_v1 = client.CoreV1Api()
    networking_v1 = client.NetworkingV1Api()

    # Example WebApp
    webapp = WebApp(
        name="example-webapp",
        namespace="default",
        spec=WebAppSpec(
            image="nginx:1.14",
            port=80,
            replicas=3,
            ssl=SSLConfig(
                enabled=True,
                secret_name="webapp-tls"
            ),
            resources=ResourceRequests(
                limits=Resources(
                    cpu="500m",
                    memory="512Mi"
                ),
                requests=Resources(
                    cpu="250m",
                    memory="256Mi"
                )
            )
        )
    )

    try:
        # Create deployment
        deployment = create_deployment(webapp)
        apps_v1.create_namespaced_deployment(
            namespace=webapp.namespace,
            body=deployment
        )
        print(f"Created deployment: {webapp.name}")

        # Create service
        service = create_service(webapp)
        core_v1.create_namespaced_service(
            namespace=webapp.namespace,
            body=service
        )
        print(f"Created service: {webapp.name}")

        # Create ingress if domains are specified and SSL is enabled
        if webapp.spec.domains and webapp.spec.ssl and webapp.spec.ssl.enabled:
            ingress = create_ingress(webapp)
            networking_v1.create_namespaced_ingress(
                namespace=webapp.namespace,
                body=ingress
            )
            print(f"Created ingress: {webapp.name}")

    except client.rest.ApiException as e:
        print(f"Kubernetes API error: {e}")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()